diff --git a/mlir/lib/Dialect/QCO/Transforms/Mapping/Mapping.cpp b/mlir/lib/Dialect/QCO/Transforms/Mapping/Mapping.cpp
index ef37256..25b6757 100644
--- a/mlir/lib/Dialect/QCO/Transforms/Mapping/Mapping.cpp
+++ b/mlir/lib/Dialect/QCO/Transforms/Mapping/Mapping.cpp
@@ -1066,7 +1066,18 @@ private:
             return WalkResult::advance();
           }

-          for (const auto& [op, indices] : ready) {
+          SmallVector<Operation*> orderedOps;
+          orderedOps.reserve(ready.size());
+          for (Operation* op : ready.keys()) {
+            orderedOps.emplace_back(op);
+          }
+          llvm::sort(orderedOps, [](Operation* lhs, Operation* rhs) {
+            assert(lhs->getBlock() == rhs->getBlock());
+            return lhs->isBeforeInBlock(rhs);
+          });
+
+          for (Operation* op : orderedOps) {
+            const auto& indices = ready.find(op)->second;
             if (isa<UnitaryOpInterface>(op)) {
               const auto i0 = indices[0];
               const auto i1 = indices[1];
